idx/memory/tag_query.go

```diff
@@ -0,0 +1,404 @@
+			case NOT_MATCH:
+				q.notMatch = append(q.notMatch, e.kv)
+			}
+		}
+	}
+
+	return q, nil
+}
+
+func (q *TagQuery) getInitialByMatch(index TagIndex) (int, TagIDs, error) {
+	resultSet := make(TagIDs)
+	lowestCount := math.MaxUint32
+	startId := 0
+
+	// choose key that has the smallest number of values because we want to
+	// reduce the number of matches and start with the smallest possible resultSet.
+	// the smaller the resultSet the less time will have to be spent to further
+	// filter its values in the following expressions.
+	for i := range q.match {
+		l := len(index[q.match[i].key])
+		if l < lowestCount {
+			lowestCount = l
+			startId = i
+		}
+	}
+
+	// shortcut if pattern is ^.+ (f.e. expression "key!=~" will be translated to "key=~^.+")
+	if q.match[startId].value == "^.+" {
+		for _, ids := range index[q.match[startId].key] {
+			for id := range ids {
+				resultSet[id] = struct{}{}
+			}
+		}
+		return startId, resultSet, nil
+	}
+
+	re, err := regexp.Compile(q.match[startId].value)
+	if err != nil {
+		return 0, nil, errInvalidQuery
+	}
+
+	for v, ids := range index[q.match[startId].key] {
+		if !re.MatchString(v) {
+			continue
+		}
+
+		for id := range ids {
+			resultSet[id] = struct{}{}
+		}
+	}
+	return startId, resultSet, nil
+}
+
+func (q *TagQuery) getInitialByEqual(index TagIndex) (int, TagIDs) {
+	resultSet := make(TagIDs)
+	lowestCount := math.MaxUint32
+	startId := 0
+
+	// choose key-value combo that has the smallest number of series IDs because
+	// we want to start with the smallest possible resultSet.
+	// the smaller the resultSet the less time will have to be spent to further
+	// filter its values in the following expressions.
+	for i := range q.equal {
+		l := len(index[q.equal[i].key][q.equal[i].value])
+		if l < lowestCount {
+			lowestCount = l
+			startId = i
+		}
+	}
+
+	// copy the map, because we'll later delete items from it
+	for k, v := range index[q.equal[startId].key][q.equal[startId].value] {
+		resultSet[k] = v
+	}
+
+	return startId, resultSet
+}
+
+// filterByEqual filters a list of metric ids by the given expressions. if "not"
+// is true it will remove all ids of which at least one expression is equal to
+// one of its tags. if "not" is false the functionality is inverted, so it
+// removes all the ids where no tag is equal to at least one of the expressions.
+//
+// expressions: the list of key & value pairs
+// skipEqual:   an integer specifying an expression index that should be skipped
+// resultSet:   list of series IDs that should be filtered
+// index:       the tag index based on which to evaluate the conditions
+// not:         whether the resultSet shall be filtered by a = or != condition
+//
+func (q *TagQuery) filterByEqual(expressions []kv, skipEqual int, resultSet TagIDs, index TagIndex, not bool) {
+	for i, e := range expressions {
+		if i == skipEqual {
+			continue
+		}
+
+		for id := range resultSet {
+			if _, ok := index[e.key][e.value][id]; ok == not {
```

Comment:
Can we hold a reference to `index[e.key][e.value]` to avoid the repeated lookups (if the `resultSet` is large this may be worthwhile).

Reply:
https://github.com/grafana/metrictank/commit/3fed95307b54f26454f61d9fb7fb264cf6772c5c